### PR TITLE
[C-1628] Add metamask/rainbow mobile wallet-connect

### PR DESCRIPTION
**Pull request by dylanjeffers**
### Description

Adds wallet flow for wallet-connect, implementing metamask and rainbow wallet

### Dragons

The layout of the code is pretty confusing, with various handoffs between component and sagas, but this is the cleanest way i could come up with now avoiding a bigger refactor. Thinking we get solana-phone implementation done next, and then consider a pattern that makes sense for the group.

There is a timeout for connecting wallet-connect wallets between connecting and signing. There seems to be a bit of inconsistency here with timing. I'll be on the lookout for a better way to handle this moving forward, whether its reducing the timeout, or waiting for a "session connected" event.

https://user-images.githubusercontent.com/8230000/206570887-917aea4e-2f1d-4811-a1fb-7e377a1af88b.MP4

### CODE REVIEW

File: packages/mobile/src/screens/wallet-connect/WalletConnectStack.tsx [hunk 1]
```diff
@@ -20,12 +20,15 @@ export const WalletConnectStack = () => {
...
       <WalletConnectProvider
-        redirectUrl='audius://'
+        redirectUrl='audius://wallets'
         storageOptions={{
           // @ts-ignore: IAsyncStorage isn't up to date
           asyncStorage: AsyncStorage
         }}
-        renderQrcodeModal={WalletConnectProviderRenderModal}
+        renderQrcodeModal={(props) => {
+          console.log('qr props', props)
```
Comment (sliptype) [inline packages/mobile/src/screens/wallet-connect/WalletConnectStack.tsx:29]: Log

File: packages/mobile/src/screens/wallet-connect/components/WalletConnectDrawer.tsx [hunk 2]
```diff
@@ -85,23 +91,27 @@ export const WalletConnectDrawer = () => {
...
+      )
     }
-  }, [visible, dispatch])
+  }, [visible, dispatch, uri])
 
   // When the drawer gets dismissed, dismiss the wallet connect popup
   useEffect(() => {
     if (visible && !isDrawerVisible) {
-      onDismiss()
+      // onDismiss()
```
Comment (sliptype) [inline packages/mobile/src/screens/wallet-connect/components/WalletConnectDrawer.tsx:112]: 👀

File: packages/mobile/src/store/wallet-connect/sagas/connectNewWalletSaga.ts [hunk 3]
```diff
@@ -97,11 +104,30 @@ function* connectNewWalletAsync(action: ConnectNewWalletAction) {
       break
     }
     case 'wallet-connect': {
-      const { connector } = action.payload
-      if (!connector) return
+      const { publicKey } = action.payload
+      // console.log('connected?', connector.session.connected)
+      // const accountKey = connector?.session.accounts[0]
```
Comment (sliptype) [inline packages/mobile/src/store/wallet-connect/sagas/connectNewWalletSaga.ts:109]: Comments

File: packages/mobile/src/store/wallet-connect/sagas/signMessageSaga.ts [hunk 4]
```diff
@@ -45,18 +43,21 @@ function* signMessageAsync(action: SignMessageAction) {
       break
     }
     case 'solana-phone-wallet-adapter': {
-      const signature = data
-      if (!publicKey) return
+      const { data: signature, publicKey } = action.payload
       // TODO: Refactor token-dashboard sagas #
```
Comment (sliptype) [inline packages/mobile/src/store/wallet-connect/sagas/signMessageSaga.ts:47]: Is this todo done?

File: packages/mobile/src/store/wallet-connect/sagas/signMessageSaga.ts [hunk 5]
```diff
@@ -45,18 +43,21 @@ function* signMessageAsync(action: SignMessageAction) {
...
     }
     case 'wallet-connect': {
-      const signature = data
-      // TODO: Refactor token-dashboard sagas #
-      // yield* put(addNewWallet({ signature, publicKey }))
-      console.log('signedMessagPayload', signature, publicKey)
+      const { data: signature } = action.payload
+
+      const updatedUserMetadata = yield* associateNewWallet(signature)
+      console.log({ updatedUserMetadata })
```
Comment (sliptype) [inline packages/mobile/src/store/wallet-connect/sagas/signMessageSaga.ts:56]: More logs and comments here

**Review by sliptype — APPROVED** [review]
I don't quite understand how this all fits together but the code looks good to me

**Comment by dylanjeffers** [comment]
@sliptype added a video to help with understanding flow!

File: packages/mobile/src/store/wallet-connect/sagas/signMessageSaga.ts (hunk 4 above)
Comment (dylanjeffers) [reply packages/mobile/src/store/wallet-connect/sagas/signMessageSaga.ts:47]: yeah potentially, not sure of exact status with solana-phone, but will be addressed in next pr

**Comment by sliptype** [comment]
> @sliptype added a video to help with understanding flow!

Oooo this is so sick! I assume if you close the drawer while the spinner is going it will finish up the the background

**Comment by dylanjeffers** [comment]
> > @sliptype added a video to help with understanding flow!
> 
> Oooo this is so sick! I assume if you close the drawer while the spinner is going it will finish up the the background

yeah maybe not rn, but def the plan :D

**Pull request merged**